### ISSUE TITLE
Project3D class

### ISSUE THREAD

**Issue by petroskey** [nianticlabs/monodepth2#52]
Hi, thank you always a lot to advice my question!
I have a question about Project3D class, wherein

pix_coords[..., 0] /= self.width - 1
pix_coords[..., 1] /= self.height - 1
pix_coords = (pix_coords - 0.5) * 2

does what? I guess that pix_coords have the projected coordinates of the temporal- or stereo-pair image plane retrieved from estimated 3d point cloud.

Thank you in advance :)

**Comment by daniyar-niantic** [comment]
Hi @petroskey 

The result of Project3D is used in [trainer.py](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L381) by `torch.nn.functional.grid_sample` which assumes that the sampling coordinates are in range [-1, 1]. Hence, the three lines of code that you are asking about are just rescaling the pixel coordinates to [-1, 1] range.

**Comment by petroskey** [comment]
Hi @daniyar-niantic,
Thank you fever much for your quick response! I see what you say:)
Another problem is that I cannot understand why this rescaling is required to compute consistency loss. It seems that the original reprojected image without rescaling can be used to compare against the input image. 
Thank you.

**Comment by mdfirman** [comment]
Hi @petroskey ,

By "this rescaling" do you mean the `pix_coords[..., 0] /= self.width - 1` rescaling in the Project3D class that you referred to above?

**Comment by petroskey** [comment]
Hi @mdfirman,
Thank you for your reply! Yes, I mean the lines
pix_coords[..., 0] /= self.width - 1
pix_coords[..., 1] /= self.height - 1
pix_coords = (pix_coords - 0.5) * 2
in the Project3D class.

I have interpreted that in your code the trainer firstly estimates the disparity or depth map from an input image, then reprojects the 3d point clouds to the pair image using the camera parameters. Is this right? To my understanding, the loss function measures inconsistency between the reprojected image and its corresponding original image. So I cannot understand why such a rescaling is needed.

Thank you always for your continuous support.

**Comment by mdfirman** [comment]
Ok, I understand. 

> I have interpreted that in your code the trainer firstly estimates the disparity or depth map from an input image, then reprojects the 3d point clouds to the pair image using the camera parameters. Is this right? 

Yes.

As @daniyar-niantic says, the rescaling code is needed to ensure that the input to `grid_sample` is scaled correctly. From the `grid_sample` documentation that was linked to:

> grid specifies the sampling pixel locations normalized by the input spatial dimensions. Therefore, it should have most values in the range of [-1, 1]. For example, values x = -1, y = -1 is the left-top pixel of input, and values x = 1, y = 1 is the right-bottom pixel of input.

If we didn't do the rescaling, then the coordinates would be in the range `[0, image_width]`, but we want them to be in the range `[-1, 1]`. The rescaling is performed to linearly transform the coordinates from the range `[0, image_width]` to the range `[-1, 1]`. This ensures that the sampled pixels are sampled  from the correct location in the source image.

One option to help you to understand this code is to comment it out, and to see what effect its removal has on the training process.

Thanks

**Comment by petroskey** [comment]
Thank you @mdfirman,
I understand that the rescaling is needed for the sampled points to be applied to grid_sample function.
Thank you @daniyar-niantic and @mdfirman for your kind reply:) It was a great help for me!!

**Comment by cswwp** [comment]
> Ok, I understand.
> 
> > I have interpreted that in your code the trainer firstly estimates the disparity or depth map from an input image, then reprojects the 3d point clouds to the pair image using the camera parameters. Is this right?
> 
> Yes.
> 
> As @daniyar-niantic says, the rescaling code is needed to ensure that the input to `grid_sample` is scaled correctly. From the `grid_sample` documentation that was linked to:
> 
> > grid specifies the sampling pixel locations normalized by the input spatial dimensions. Therefore, it should have most values in the range of [-1, 1]. For example, values x = -1, y = -1 is the left-top pixel of input, and values x = 1, y = 1 is the right-bottom pixel of input.
> 
> If we didn't do the rescaling, then the coordinates would be in the range `[0, image_width - 1]`, but we want them to be in the range `[-1, 1]`. The rescaling is performed to linearly transform the coordinates from the range `[0, image_width - 1]` to the range `[-1, 1]`. This ensures that the sampled pixels are sampled from the correct location in the source image.
> 
> One option to help you to understand this code is to comment it out, and to see what effect its removal has on the training process.
> 
> Thanks

Hi @mdfirman @daniyar-niantic  I am confused by the constructed loss. In my think, 
-- use_stereo true
1: left image + network to get the disparity
2: disparity + camera params to get point cloud
3: point cloud  + camera intrinsice, extrinsics  to get the other side rgb image
Is my think right?
If yes, i have one question:
i notice that the compute_reprojection_loss is between different side as the image shows, and i can't understand, Do you mind to explain?
![image](https://user-images.githubusercontent.com/11901683/78887436-70880e00-7a92-11ea-8eba-f81bee695541.png)
